File: LowLevelDesign/observer-pattern/pubsub/src/topic.ts
import { Message } from "./message";
import { ISubscriber } from "./subscriber";

export class Topic {
    private readonly _name: string;
    private readonly _subscribers: Set<ISubscriber>;

    constructor(name: string) {
        this._name = name;
        this._subscribers = new Set();
    }

    addSubscriber(subscriber: ISubscriber): void {
        this._subscribers.add(subscriber)
    }

    removeSubscriber(subscriber: ISubscriber): void {
        this._subscribers.delete(subscriber)
    }

    publish(message: Message): void {
        for (const subscriber of this._subscribers) {
            subscriber.onMessage(this._name, message.content);
        }
    }
}

File: SystemDesignDiagrams/practice/LLD/PubSub/src/main/java/com/pubsub/entities/Topic.java
package com.pubsub.entities;

import com.pubsub.subscriber.Subscriber;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Topic {
    private final String name;
    private Set<Subscriber> subscribers = new HashSet<>();
    private final ExecutorService deliveryExecutor;

    public Topic(String topicName, ExecutorService deliveryExecutor) {
        this.name = topicName;
        this.deliveryExecutor = deliveryExecutor;
    }

    public String getName() {
        return name;
    }

    public void addSubscriber(Subscriber subscriber) {
        subscribers.add(subscriber);
    }

    public void removeSubscriber(Subscriber subscriber) {
        subscribers.remove(subscriber);
    }

    public void notify(Message message) {
        for(Subscriber subscriber: subscribers) {
            deliveryExecutor.submit(() -> {
                try {
                    subscriber.onUpdate(this, message);
                } catch(Exception e) {
                    System.err.println("Error while delivering message to subscriber " + e.getMessage());
                }
            });
        }
    }
}
